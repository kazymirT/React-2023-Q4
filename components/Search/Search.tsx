import React, { useRef } from 'react';
import styles from './styles.module.css'
import { useCallback, useEffect, useState } from 'react';
import { updateSearchParams } from '../utils/updateSearchParams';
import { useRouter } from 'next/router';

export const Search = () => {
  const router = useRouter();
  
  const handlerSubmit = useCallback(
    (event: React.ChangeEvent<HTMLFormElement>) => {
      event.preventDefault();
      const form = new FormData(event.target)
      const searchValue: string | null = form.get('search') as string;
      const newUrl = updateSearchParams('search', searchValue);
      router.push(newUrl);
    },
    [router]
  );

  return (
      <div className={styles.search}>
        <form role="search" onSubmit={handlerSubmit}>
          <input 
            className={styles.input}
            type="text"
            name="search"
            defaultValue={router.query.search}
            placeholder="What product are you looking for?"
          />
          <button className={styles.btn} type="submit" id="search-button">
            Search
          </button>
        </form>
      </div>
  );
};
