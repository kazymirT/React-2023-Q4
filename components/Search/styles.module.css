.search {
  padding: 40px 0;
  width: 100%;
  max-width: 1280px;
  margin: 0 auto;
  display: flex;
  justify-content: center;
};

.btn {
  display: block;
  width: 100px;
  margin-right: 5px;
  padding: 10px 15px;
  border-radius: 10px;
  cursor: pointer;
  background-color: yellow;
}
.btn:hover {
  background-color: lightcoral;
}
.input {
  width: 300px;
  padding: 10px 15px;
  margin-right: 25px;
  border-radius: 10px;
  outline: none;
  border: 5px solid;
  border-color: green;
}
