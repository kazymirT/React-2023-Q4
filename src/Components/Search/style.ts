import styled from 'styled-components';

export const SearchDiv = styled.div`
  padding: 40px 0;
  width: 100%;
  max-width: 1280px;
  margin: 0 auto;
  display: flex;
  justify-content: center;
`;

export const Button = styled.button`
  width: 100px;
  margin-right: 5px;
  padding: 10px 15px;
  border-radius: 10px;
  cursor: pointer;
  background-color: ${(props) => (props.type === 'reset' ? 'yellow' : 'red')};
  &:hover {
    background-color: ${(props) =>
      props.type === 'reset' ? 'lightyellow' : 'lightcoral'};
  }
`;

export const Input = styled.input`
  width: 300px;
  padding: 10px 15px;
  margin-right: 25px;
  border-radius: 10px;
  outline: none;
  border: 5px solid;
  border-color: ${(props) =>
    typeof props.value === 'string' && props.value.length > 0
      ? 'green'
      : 'gray'};
`;
