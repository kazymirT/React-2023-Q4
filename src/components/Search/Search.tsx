import React from "react";
import styles from "./styles.module.css";
import { useCallback } from "react";
import { updateSearchParams } from "../utils/updateSearchParams";
import { useRouter } from "next/router";

export const Search = () => {
  const router = useRouter();
  const handlerSubmit = useCallback(
    (event: React.ChangeEvent<HTMLFormElement>) => {
      event.preventDefault();
      const form = new FormData(event.target);
      const searchValue: string | null = form.get("search") as string;
      const newUrl = updateSearchParams("search", searchValue);
      localStorage.setItem("search", searchValue);
      router.push(newUrl);
    },
    [router],
  );

  return (
    <div className={styles.search}>
      <form role="search" data-testid={"search-form"} onSubmit={handlerSubmit}>
        <input
          className={styles.input}
          data-testid={"search-input"}
          type="text"
          name="search"
          defaultValue={router.query.search}
          placeholder="What product are you looking for?"
        />
        <button className={styles.btn} type="submit" id="search-button">
          Search
        </button>
      </form>
    </div>
  );
};
