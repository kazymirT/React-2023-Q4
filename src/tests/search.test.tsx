import React from "react";
import { describe, expect, it, vi } from "vitest";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import mockRouter from "next-router-mock";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";
import { Search } from "@/components/Search/Search";

vi.mock("next/router", () => vi.importActual("next-router-mock"));

describe("Tests for the Search component", () => {
  it("Clicking the Search button saves the entered value to the local storage", async () => {
    mockRouter.push("/");
    render(<Search />, {
      wrapper: MemoryRouterProvider,
    });
    expect(localStorage.getItem("search")).toBe(null);
    fireEvent.change(screen.getByTestId("search-input"), {
      target: { value: "phone" },
    });
    fireEvent.submit(screen.getByTestId("search-form"));

    expect(mockRouter.asPath).toBe("/?search=phone&page=1");
    expect(localStorage.getItem("search")).toBe("phone");

    fireEvent.change(screen.getByTestId("search-input"), {
      target: { value: "watch" },
    });
    fireEvent.submit(screen.getByTestId("search-form"));
    expect(screen.getByTestId("search-form")).toMatchSnapshot();

    expect(mockRouter.asPath).toBe("/?search=watch&page=1");
    expect(localStorage.getItem("search")).toBe("watch");
  });

  it("The component retrieves the value from the local storage upon mounting.", async () => {
    const localStorageValue = localStorage.getItem("search");
    expect(localStorageValue).toBe("watch");
    mockRouter.push(`/?search=${localStorageValue}`);
    render(<Search />, {
      wrapper: MemoryRouterProvider,
    });

    expect(screen.getByTestId("search-form")).toMatchSnapshot();
    expect(screen.getByTestId("search-input")).toHaveValue(localStorageValue);
  });
});
