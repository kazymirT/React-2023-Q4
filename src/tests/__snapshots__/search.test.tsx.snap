// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Search component > Clicking the Search button saves the entered value to the local storage 1`] = `
<form
  data-testid="search-form"
  role="search"
>
  <input
    class="_input_21ed85"
    data-testid="search-input"
    name="search"
    placeholder="What product are you looking for?"
    type="text"
    value="watch"
  />
  <button
    class="_btn_21ed85"
    id="search-button"
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Tests for the Search component > The component retrieves the value from the local storage upon mounting. 1`] = `
<form
  data-testid="search-form"
  role="search"
>
  <input
    class="_input_21ed85"
    data-testid="search-input"
    name="search"
    placeholder="What product are you looking for?"
    type="text"
    value="watch"
  />
  <button
    class="_btn_21ed85"
    id="search-button"
    type="submit"
  >
    Search
  </button>
</form>
`;
